### ISSUE TITLE
Support activity/revisions/logs retention

### ISSUE THREAD

**Issue by rijkvanzanten** [directus/directus#23166]
We need to add a set of new environment variables to control the maximum retention of activity/revision/flow logs rows:

`ACTIVITY_RETENTION`
`REVISIONS_RETENTION`
`FLOWS_LOG_RETENTION`

These will be set in a time value (eg `30d`, `2m`, just like `ACCESS_TOKEN_TTL`). 

On a cron job (set through `RETENTION_CRON`) we need to have one connected process (make sure to add a lock) delete the old items in batches (set through `RETENTION_BATCH`). To do this, in a while loop, we should do a `SELECT LIMIT 1` to check if there's anything to delete. If that returns something, we should run a delete using a subquery for a selection, for example:

```sql
DELETE FROM directus_activity WHERE id IN (
    SELECT id 
    FROM directus_activity
    WHERE timestamp > {{ today - $ACTIVITY_RETENTION }}
    LIMIT {{ $RETENTION_BATCH }}
);
```

We should use an index (pending #23149) on `timestamp` to speed up this subquery. The limit is important to not block up the database during these loops.

In the migration to add the index, we have to make sure we run the deletion once _first_ as adding a new index on a giant table can take a while and blocks up the database. By doing a slower batch delete first we can ensure the index can be created more swiftly afterwards.

**Comment by rijkvanzanten** [comment]
Also see linked issue above (https://github.com/directus/directus/issues/17894) with more context around the problems we're solving.

**Comment by ComfortablyCoding** [comment]
This is also blocked by https://github.com/directus/directus/pull/22413 and https://github.com/directus/directus/pull/22295

**Comment by msbt** [comment]
Hey there! Is there an eta on the corresponding PR https://github.com/directus/directus/pull/23410? Like a poster in a different issue mentioned, I also have a rather small instance, but the activity, flows and revisions are ballooning it up beyond 15GB, which brings me on the verge of rescaling the machine even though there would be no need if the logs got pruned at some point. 

Is there a quick & safe fix to do this manually so I don't have to scale before the PR gets merged?

**Comment by rijkvanzanten** [comment]
@msbt If you don't care about retaining the old activity/revisions at all, a TRUNCATE is really fast and will work just fine. If you want to retain a subset of those records, it gets a bit more complicated as databases can slow down quite a bit when doing operations against a large table; hence the various PRs around the subject

**Comment by msbt** [comment]
@rijkvanzanten thanks, I'll investigate :)

**Comment by jclaveau** [comment]
> @rijkvanzanten thanks, I'll investigate :)

To make the `delete from directus_activity` fast you need an index on the `directus_revisions.activity`  like indicated here https://dba.stackexchange.com/a/138270/203054

With it I've been able to run requests like `delete from directus_activity where timestamp < 'xxxx'::date` in less than 2 minutes on millions of records

**Comment by Adashi12** [comment]
Maybe this is of some interest here, haven't tested on a recent version though:
[https://gist.github.com/lukas-schaetzle/f93eff3d961ac595d0e1ab4be5f34536](https://gist.github.com/lukas-schaetzle/f93eff3d961ac595d0e1ab4be5f34536)

**Comment by msbt** [comment]
@Adashi12 thanks, I was using a modified version of that for postgres and gained 15GB of space :pray:

**Comment by rijkvanzanten** [comment]
@jclaveau Yup! But that initial migration can take a while and even timeout for bigger existing projects. Strategy here is to first split it up into smaller chunks (eg comments https://github.com/directus/directus/pull/22295 / versions https://github.com/directus/directus/pull/22413 to a different table), then to add the necessary indexes (#23149) to activity/revisions and the other new tables, and then to enable the retention control for projects (https://github.com/directus/directus/pull/23410) 🙂

**Comment by jclaveau** [comment]
> @jclaveau Yup! But that initial migration can take a while and even timeout for bigger existing projects. Strategy here is to first split it up into smaller chunks (eg comments #22295 / versions #22413 to a different table), then to add the necessary indexes (#23149) to activity/revisions and the other new tables, and then to enable the retention control for projects (#23410) 🙂


I'm happy I handled it before requiring this work! 

Also, did you consider storing the `user_agent` field of `directus_activity` and `directus_session` in a dedicated table. I made the calculation a while ago and it used respectively 44% and 32% of their disk usage on Postgres (The same calculation gave respectively 5.2% and 3.9% for the `ip` fields btw)?

**Comment by rijkvanzanten** [comment]
> I'm happy I handled it before requiring this work!

😄  Luckily this stuff all happens under the hood and in migrations so the end user shouldn't notice a thing (except for the tables to get way smaller of course).

> Also, did you consider storing the user_agent field of directus_activity and directus_session in a dedicated table. I made the calculation a while ago and it used respectively 44% and 32% of their disk usage on Postgres (The same calculation gave respectively 5.2% and 3.9% for the ip fields btw)?

I have not, but that's a great call! UAs can be quite long and in the context of Activity are often duplicated. Makes me wonder if we should do the same for IP addresses, although those aren't as large of course. There's also a more aggressive question of _Do we even care about the UA in an activity log?_

**Comment by jclaveau** [comment]
> Do we even care about the UA in an activity log?

Good point! 
- A big part of this value is already handled by Google Analytics or other for free 
- Maybe being able to pipe it to some dedicated infra like elastic search would be more consistent (non relational db seems naturally more adapted to events analytics)
- also some dedicated lib may reduce the stored data by analyzing it first like https://github.com/faisalman/ua-parser-js (which still let factorizing the data relevant)
